pkg/cvo/upgradeable.go

```diff
@@ -234,6 +247,128 @@ func (check *clusterManifestDeleteInProgressUpgradeable) Check() *configv1.Clust
+	if ackConfigmapErr != nil {
+		if apierrors.IsNotFound(ackConfigmapErr) {
+			errorMessage = strings.Trim(errorMessage+" "+fmt.Sprintf("%s configmap not found.", internal.AdminAcksConfigMap), " ")
+		} else if ackConfigmapErr != nil {
+			errorMessage = strings.Trim(errorMessage+" "+
+				fmt.Sprintf("Unable to access configmap %s, err=%v.", internal.AdminAcksConfigMap, ackConfigmapErr), " ")
+		}
+	}
+	if errorMessage != "" {
+		klog.Error(errorMessage)
+		if gateMessage != "" {
+			errorMessage = gateMessage + " " + errorMessage
+		}
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "AdminAckConfigMapError",
+			Message: errorMessage,
+		}
+	}
+	if val, ok := ackConfigmap.Data[gate]; !ok || val != "true" {
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "AdminAckRequired",
+			Message: gateMessage,
+		}
+	}
+	return nil
+}
+
+type clusterAdminAcksCompletedUpgradeable struct {
+	adminGatesLister listerscorev1.ConfigMapNamespaceLister
+	adminAcksLister  listerscorev1.ConfigMapNamespaceLister
+	cvLister         configlistersv1.ClusterVersionLister
+	cvoName          string
+}
+
+func (check *clusterAdminAcksCompletedUpgradeable) Check() *configv1.ClusterOperatorStatusCondition {
+	cv, err := check.cvLister.Get(check.cvoName)
+	if meta.IsNoMatchError(err) || apierrors.IsNotFound(err) {
+		message := fmt.Sprintf("Unable to get ClusterVersion, err=%v", err)
+		klog.Error(message)
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "UnableToGetClusterVersion",
+			Message: message,
+		}
+	}
+	currentVersion := clusterversion.GetCurrentVersion(cv.Status.History)
+
+	// This can occur in early start up when the configmap is first added and version history
+	// has not yet been populated.
+	if currentVersion == "" {
+		return nil
+	}
+
+	var gateCm *corev1.ConfigMap
+	if gateCm, err = check.adminGatesLister.Get(internal.AdminGatesConfigMap); err != nil {
+		var message string
+		if apierrors.IsNotFound(err) {
+			message = fmt.Sprintf("%s configmap not found", internal.AdminGatesConfigMap)
+		} else if err != nil {
+			message = fmt.Sprintf("Unable to access configmap %s, err=%v", internal.AdminGatesConfigMap, err)
+		}
+		klog.Error(message)
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "UnableToAccessAdminGatesConfigMap",
+			Message: message,
+		}
+	}
+	ackCm, err := check.adminAcksLister.Get(internal.AdminAcksConfigMap)
+	return checkAdminGate("ack-4.8-kube-122-api-removals-in-4.9", currentVersion, gateCm, ackCm, err)
```

Comment:
Hmm.  Are we hard-coding these keys here to avoid the risk of losing a race between admin removing a key from the gate ConfigMap and the manifest-reconciliation loop noticing and stuffing that key back in?

Reply:
We hard-coded to avoid added complexity of making them configurable per enhancement.

Reply:
I wasn't expecting to have the gates hardcoded here, if we add another we'd have to update this code?
I thought we were going to match `ack-4.8` against the current major.minor and anything that matched would get checked for a corresponding admin ack.
We won't need a new enhancement for each gate, we filed an enhancement for the general pattern with a specific example for context.

Reply:
Sorry, hard-coded means "in the code" to me but I put a hold on this. So anything added to the gates cm, along as it fits the required naming pattern, is valid? And if it's not in the ack cm Upgradeable=False.

Reply:
Re-worked to remove the hard-coded gate names from CVO go code.